dhis-2/dhis-services/dhis-service-tracker/src/main/java/org/hisp/dhis/tracker/export/trackedentity/DefaultTrackedEntityService.java

```diff
@@ -97,10 +105,91 @@ class DefaultTrackedEntityService implements TrackedEntityService {
 
   private final EventService eventService;
 
+  private final FileResourceService fileResourceService;
+
   private final TrackedEntityOperationParamsMapper mapper;
 
   private final UserService userService;
 
+  @Override
+  public FileResourceStream getFileResource(
+      UID trackedEntity, UID attribute, @CheckForNull UID program) throws NotFoundException {
+    FileResource fileResource = getFileResourceMetadata(trackedEntity, attribute, program);
+
+    return new FileResourceStream(
+        fileResource,
+        () -> {
+          try {
+            return fileResourceService.openContentStream(fileResource);
+          } catch (NoSuchElementException e) {
+            // Note: we are assuming that the file resource is not available yet. The same approach
+            // is taken in other file endpoints or code relying on the storageStatus = PENDING.
+            // All we know for sure is the file resource is in the DB but not in the store.
+            throw new ConflictException(
+                "The content is being processed and is not available yet. Try again later.");
+          } catch (IOException e) {
+            throw new ConflictException(
+                "Failed fetching the file from storage",
+                "There was an exception when trying to fetch the file from the storage backend. "
+                    + "Depending on the provider the root cause could be network or file system related.");
+          }
+        });
+  }
+
+  private FileResource getFileResourceMetadata(
+      UID trackedEntityUid, UID attributeUid, @CheckForNull UID programUid)
+      throws NotFoundException {
+    TrackedEntity trackedEntity = trackedEntityStore.getByUid(trackedEntityUid.getValue());
+    if (trackedEntity == null) {
+      throw new NotFoundException(TrackedEntity.class, trackedEntityUid.getValue());
+    }
+
+    TrackedEntityAttribute attribute =
+        trackedEntityAttributeService.getTrackedEntityAttribute(attributeUid.getValue());
+    if (attribute == null) {
+      throw new NotFoundException(TrackedEntityAttribute.class, attributeUid.getValue());
+    }
+
+    if (!attribute.getValueType().isFile()) {
+      throw new NotFoundException(
+          "Tracked entity attribute " + attributeUid.getValue() + " is not a file (or image).");
+    }
+
+    List<String> errors;
+    User currentUser = userService.getUserByUsername(CurrentUserUtil.getCurrentUsername());
+    if (programUid != null) {
+      Program program = programService.getProgram(programUid.getValue());
+      if (program == null) {
+        throw new NotFoundException(Program.class, programUid.getValue());
+      }
+
+      errors = trackerAccessManager.canRead(currentUser, trackedEntity, program, false);
+    } else {
+      errors = trackerAccessManager.canRead(currentUser, trackedEntity);
```

Comment:
is this check necessary? is it not already done when getting the tracked entity in line 142?

Reply:
It is AFAIK

https://github.com/dhis2/dhis2-core/blob/4d34f1d4098447d2cfdc48752202dd21ef2a1dcd/dhis-2/dhis-services/dhis-service-core/src/main/java/org/hisp/dhis/trackedentity/DefaultTrackerAccessManager.java#L63

is doing much more.

The generic stores backing things like `trackedEntityStore.getByUid` only check the sharing of the entity itself.